tests/unit_tests/steps/test_analyse.py

```diff
@@ -1,22 +1,139 @@
 from pathlib import Path
+from unittest import mock
+from unittest.mock import mock_open
 
 import pytest
+from fab.util import HashedFile
 
 from fab.dep_tree import AnalysedFile
 from fab.steps.analyse import Analyse
 
 
+# These methods are just glue code and do not currently have a test.
+#   _analyse_source_code
+#   _analyse_dependencies
+#   _parse_files
+
+
+@pytest.fixture
+def analyser():
+    analyser = Analyse()
+    analyser._config = mock.Mock(multiprocessing=False, project_workspace=Path('workspace'))
+    return analyser
+
+
+def test_get_file_checksums(analyser):
+    # ensure this function returns the expected data structure
+
+    fpaths = [Path('foo.c'), Path('bar.c')]
+
+    def fake_hasher(fpath):
+        return HashedFile(fpath, hash(str(fpath)))
+
+    # we don't actually try to hash the files in this test, we just return something repeatable as a fake hash
+    with mock.patch('fab.steps.analyse.do_checksum', side_effect=fake_hasher):
+        result = analyser._get_file_checksums(fpaths)
+
+    expected = {fpath: hash(str(fpath)) for fpath in fpaths}
+    assert result == expected
+
+
+class Test_load_analysis_results(object):
+
+    @pytest.fixture
+    def csv_lines(self):
+        # a module with a dependecy on a fortran and c file, plus a mo commented dep
+        return [
+            'fpath,file_hash,symbol_defs,symbol_deps,file_deps,mo_commented_file_deps',
+            'my_mod.f90,123,my_mod,dep1_mod;dep2,,mo_dep.c',
+            'dep1_mod.f90,234,dep1_mod,,,',
+            'dep2.c,345,dep2,,,',
+        ]
+
+    @pytest.fixture
+    def latest_file_hashes(self):
+        return {Path('my_mod.f90'): 123, Path('dep1_mod.f90'): 234, Path('dep2.c'): 345}
+
+    def test_no_analysis_file(self, analyser):
+        # there's nothing to load
+        with mock.patch('fab.steps.analyse.open', side_effect=FileNotFoundError('mwah-ha-ha-haa')):
+            results = analyser._load_analysis_results(latest_file_hashes=dict())
+
+        assert results == dict()
+
+    def test_nothing_changed(self, analyser, csv_lines, latest_file_hashes):
```

Comment:
Can you confirm that this is testing for the correct thing? That is to say, a Fortran object file depends on the `.f90` file for its own source but the `.mod` files (not the `.o` files) of its prerequisites. This is not immediately obvious here.

Reply:
Good question. When you first raised this point about Fortran dependencies, we hadn't realised it (thank you). We see it as a key requirement for the upcoming incremental build work but Fab doesn't know it, or need to, until that work takes place.

Until then, changes to a source file can only indicate that _it_ needs to be reprocessed. The csv data in this test - which could probably be more accessibly presented - says _my_mod.f90_ depends on the symbol _dep1_mod_. This dependency is only currently used to determine compile order. With the incremental build work (pending proper requirements analysis) this will also come to indicate a dependency on _dep1_mod.mod_ and the upstream implications of that.